app/controllers/application_controller.rb

```
@@ -17,6 +17,8 @@ class ApplicationController < ActionController::Base
   # reason to call authorize in devise or CAS controller acitons.
   after_action :verify_authorized, unless: :devise_controller?
 
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
```

I'm a bit worried about this. It might be fine in non-Highlander since there is nothing other than the CAS SSO stuff on platform for non-admins. But in Highlander, this is likely going to get us in trouble when the authorization failed for some other reason and we end up in an infinite loop of sorts. Also, it will mask the actual errors from our logs and Honeycomb to be able to troubleshoot.

How about we just do this sort of thing in the `home_controller`? I think we've probably already mostly solved the issue by switching the `new_user_session_path` to `cas_login_url`, but having the redirect only on the home controller feels like a safer fallback for whatever root cause we may have missed. E.g. does password reset end up causing this and we need to switch a URL there as well?

When you implement the redirect in the `home_controller` I would explicitly check if the user has an admin role and redirect to Canvas if not instead of relying on an exception.

---

+1

---

One more potential root cause I just remembered. I think log-out can cause this b/c it log's you out to the platform log-in page but loses the `service` URL that tells login to bring you to Canvas. The above solution with the `home_controller` should fix that scenario, but in a hacky way. The proper fix is probably to preserve the `service` URL. This wasn't as much of a problem before b/c the Join server would * mostly * be able to figure out where you should go and redirect you but now with it largely out of the picture everyone who logs out and back in is perhaps ending up at the "something went wrong screen"

---

I moved the rescue from the application controller and instead ensured the user is an admin in the home controller, as requested. I created a backlog task to preserve the service url [here].(https://app.asana.com/0/1179330853755533/1199679486201904/f)